cluster-autoscaler/core/scaledown/actuation/update_latency_tracker.go

```diff
@@ -0,0 +1,117 @@
+package actuation
+
+import (
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+	"k8s.io/klog/v2"
+)
+
+const waitForTaintingTimeoutDuration = 30 * time.Second
+
+type NodeTaintStartTime struct {
+	nodeName  string
+	startTime time.Time
+}
+
+// UpdateLatencyTracker can be used to calculate round-trip time between CA and api-server
+// when adding ToBeDeletedTaint to nodes
+type UpdateLatencyTracker struct {
+	startTimestamp           map[string]time.Time
+	finishTimestamp          map[string]time.Time
+	remainingNodeCount       int
+	nodeLister               kubernetes.NodeLister
+	StartTimeChan            chan NodeTaintStartTime
+	sleepDurationWhenPolling time.Duration
+	// Passing a bool will wait for all the started nodes to get tainted and calculate
+	// latency based on latencies observed. (If all the nodes did not get tained within
+	// waitForTaintingTimeoutDuration after passing a bool, latency calculation will be
+	// aborted and the ResultChan will be closed without returning a value) Closing the
+	// AwaitOrStopChan without passing any bool will abort the latency calculation.
+	AwaitOrStopChan chan bool
+	ResultChan      chan time.Duration
+	// now is used only to make the testing easier
+	now func() time.Time
+}
+
+func NewUpdateLatencyTracker(nodeLister kubernetes.NodeLister, sleepDurationWhenPolling time.Duration, now func() time.Time) *UpdateLatencyTracker {
```

Comment:
nit: You don't need to take all these parameters here, an alternative would be to just use default and use a separate function like `NewUpdateLatencyTrackerForTesting` that uses shorter polling period and accepts a mock for `time.Now`.

Reply:
Done